### PR TITLE
[SYCL] Embedded coarse-grained debug aid for finding imbalance in Level-Zero alloc/free.

### PR DESCRIPTION
**Pull request by smaslov-intel**
A sample output is this:
```
ZE_DEBUG=4: check balance of create/destroy calls
----------------------------------------------------------
               zeContextCreate = 1     \--->              zeContextDestroy = 1
          zeCommandQueueCreate = 1     \--->         zeCommandQueueDestroy = 1
                zeModuleCreate = 1     \--->               zeModuleDestroy = 1
                zeKernelCreate = 1     \--->               zeKernelDestroy = 1
             zeEventPoolCreate = 8     \--->            zeEventPoolDestroy = 8
  zeCommandListCreateImmediate = 1     |
           zeCommandListCreate = 2     \--->          zeCommandListDestroy = 3
                 zeEventCreate = 2000  \--->                zeEventDestroy = 2000
                 zeFenceCreate = 2     \--->                zeFenceDestroy = 2
                 zeImageCreate = 0     \--->                zeImageDestroy = 0
               zeSamplerCreate = 0     \--->              zeSamplerDestroy = 0
              zeMemAllocDevice = 0     |
                zeMemAllocHost = 0     |
              zeMemAllocShared = 0     \--->                     zeMemFree = 0
```
Also fixed a bug in `piEnqueueEventsWait` where we were allocating a command-list but never closed/released it.

Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -4109,48 +4122,55 @@ pi_result piEnqueueEventsWait(pi_queue Queue, pi_uint32 NumEventsInWaitList,
   PI_ASSERT(Queue, PI_INVALID_QUEUE);
   PI_ASSERT(Event, PI_INVALID_EVENT);
 
-  _pi_ze_event_list_t TmpWaitList;
-  if (auto Res = TmpWaitList.createAndRetainPiZeEventList(NumEventsInWaitList,
-                                                          EventWaitList, Queue))
-    return Res;
+  if (EventWaitList) {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4125]: The change here is just a restructuring to put command-list allocation under this condition, i.e. when it is really needed.

Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:4125]: Thanks for the note which helped me a lot to review the change.
Nice catch here for the unnecessary command-list allocation.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -5644,6 +5672,83 @@ pi_result piTearDown(void *PluginParameter) {
   delete PiPlatformsCache;
   delete PiPlatformsCacheMutex;
 
+  // Print the balance of various create/destroy native calls.
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5675]:
```suggestion
  // Print the balance of various create/destroy native calls.
  // The idea is to verify if the number of create(+) and destroy(-) calls are matched.
```

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -5644,6 +5672,83 @@ pi_result piTearDown(void *PluginParameter) {
...
+      {"zeModuleCreate",       "zeModuleDestroy"},
+      {"zeKernelCreate",       "zeKernelDestroy"},
+      {"zeEventPoolCreate",    "zeEventPoolDestroy"},
+      {"zeCommandListCreateImmediate", "zeCommandListCreate", "zeCommandListDestroy"},
+      {"zeEventCreate",        "zeEventDestroy"},
+      {"zeFenceCreate",        "zeFenceDestroy"},
+      {"zeImageCreate",        "zeImageDestroy"},
+      {"zeSamplerCreate",      "zeSamplerDestroy"},
+      {"zeMemAllocDevice", "zeMemAllocHost", "zeMemAllocShared", "zeMemFree"},
+    };
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5696]: Do you also want to match zeDeviceGet() and zeDeviceDestroy()?

**Review by bso-intel — CHANGES_REQUESTED** [review]

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5696]: `zeDeviceGet` returns multiple devices (and they are not allocated but returned from a cache) and there is no `zeDeviceDestroy` (the devices' cache is destroyed at the end)

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -5644,6 +5672,84 @@ pi_result piTearDown(void *PluginParameter) {
   delete PiPlatformsCache;
   delete PiPlatformsCacheMutex;
 
+  // Print the balance of various create/destroy native calls.
+  // The idea is to verify if the number of create(+) and destroy(-) calls are matched.
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5676]:
```suggestion
  // The idea is to verify if the number of create(+) and destroy(-) calls are
  // matched.
```

**Review by bso-intel — APPROVED** [review]
Thanks for clarification about zeDeviceGet.
LGTM

**Comment by smaslov-intel** [comment]
@romanovvlad , @v-klochkov , @againull please merge

**Pull request merged**